### PR TITLE
Importing of BibDesk Groups and Linked Files

### LINKED ISSUE #10381: Import bib files from BIbDesk and parse the groups + linked files
**Is your suggestion for improvement related to a problem? Please describe.**
I want to switch from BibDesk (a reference manager on macOS which also uses BibTeX as basis) to JabRef and import my library in JabRef.

**Describe the solution you'd like**
I want to have the groups I used in BibDesk to be in JabRef.

Extra: Be able to parse the linked files

**Additional context**

JabRef should be able to parse the metadata and convert it to JabRef groups.

Groups In BibDesk are stored as metadata in the comment field. 
Entries in the group are identified by their citation key.

Extra: Be able to convert the linked files to format:
LInked files are stored in some kind of binary/base64 encoded file path. It's a mac os alias.

Decode the base64 data:

```
bdsk-file-1 = {YnBsaXN0MDDSAQIDBFxyZWxhdGl2ZVBhdGhZYWxpYXNEYXRhXxAbLi4vRG93bmxvYWRzLzIzMDkuMDY5MzIucGRmTxEBUgAAAAABUgACAAAMTWFjaW50b3NoIEhEAAAAAAAAAAAAAAAAAAAA4O/yLkJEAAH/////DjIzMDkuMDY5MzIucGRmAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAP/////hKRjCAAAAAAAAAAAAAQACAAAKIGN1AAAAAAAAAAAAAAAAAAlEb3dubG9hZHMAAAIAKy86VXNlcnM6Y2hyaXN0b3BoczpEb3dubG9hZHM6MjMwOS4wNjkzMi5wZGYAAA4AHgAOADIAMwAwADkALgAwADYAOQAzADIALgBwAGQAZgAPABoADABNAGEAYwBpAG4AdABvAHMAaAAgAEgARAASAClVc2Vycy9jaHJpc3RvcGhzL0Rvd25sb2Fkcy8yMzA5LjA2OTMyLnBkZgAAEwABLwAAFQACABH//wAAAAgADQAaACQAQgAAAAAAAAIBAAAAAAAAAAUAAAAAAAAAAAAAAAAAAAGY},
```

with `base64 -D | plutil -p -`
 
That gives 

```
{
  "aliasData" => {length = 338, bytes = 0x00000000 01520002 00000c4d 6163696e ... 00020011 ffff0000 }
  "relativePath" => "../Downloads/2309.06932.pdf"
}
```

Example: 
```bibtex
@comment{BibDesk Static Groups{
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<array>
	<dict>
		<key>group name</key>
		<string>testSammlung</string>
		<key>keys</key>
		<string>Kraljic:2023aa,Kovakkuni:2023aa,Pereira-Santaella:2023aa,Cooper:2023aa,Heyl:2023aa,Swain:2023aa</string>
	</dict>
</array>
</plist>
}}
```

Complete File:
[TestbIbDesk.bib.txt](https://github.com/JabRef/jabref/files/12612175/TestbIbDesk.bib.txt)

Implementation hint
Parse the metadata at 
https://github.com/JabRef/jabref/blob/2f5c51216f99e32a7453b169178564b429bdd68e/src/main/java/org/jabref/logic/importer/fileformat/BibtexParser.java#L292-L305

### PR DESCRIPTION
**Pull request by Frequinzy**
Added functionality in the BibtexParser to parse BibDesk Static Groups and convert them to Explicit Groups in JabRef.
Also added functionality for the BibtexParser to parse Linked Files in BibDesk and  convert then to Linked FIles in JabRef. 
Note: gets the relative path from the .bib file to the linked file. So if the .bib is saved in another location the link might break.

We have looked into how to implement Smart and Mixed Groups from BibDesk, but it proved difficult since they don't have any clear equivalent in JabRef. Out primary idea is that we could at least parse the Smart Group and then create an Explicit Group based on the information. This is what the added tests for Smart and Mixed Groups test for, even though that functionality isn't implemented yet.

Closes #10381

### Mandatory checks

- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

File: external-libraries.md [hunk 1]
```diff
@@ -696,26 +702,44 @@ com.github.sialcasa.mvvmFX:mvvmfx-validation:f195849ca9
...
+commons-digester:commons-digester:2.1
+commons-logging:commons-logging:1.2
+commons-validator:commons-validator:1.7
 com.oracle.ojdbc:ojdbc10:19.3.0.0
 com.oracle.ojdbc:ons:19.3.0.0
 com.oracle.ojdbc:osdt_cert:19.3.0.0
 com.oracle.ojdbc:osdt_core:19.3.0.0
 com.oracle.ojdbc:simplefan:19.3.0.0
 com.oracle.ojdbc:ucp:19.3.0.0
+compileClasspath - Compile classpath for source set
```
Comment (Siedlerchr) [inline external-libraries.md:727]: Remove this line

File: src/main/java/org/jabref/logic/importer/fileformat/BibtexParser.java [hunk 2]
```diff
@@ -330,6 +349,70 @@ private void parseJabRefComment(Map<String, String> meta) {
...
+
+    /**
+     * Adds BibDesk group entries to the JabRef database
+     * */
+    private void addBibDeskGroupEntriesToJabRefGroups() {
+        for (String groupName: parsedBibdeskGroups.keySet()) {
+            String[] citationKeys = parsedBibdeskGroups.get(groupName).split(",");
+            for (String citation : citationKeys) {
+                if (database.getEntryByCitationKey(citation).isPresent()) {
+                    String groupValue = database.getEntryByCitationKey(citation).get().getField(StandardField.GROUPS).orElse(null);
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/importer/fileformat/BibtexParser.java:365]: You can use Optionals is Present/isEmpty checks

File: src/main/java/org/jabref/logic/importer/fileformat/BibtexParser.java [hunk 3]
```diff
@@ -330,6 +349,70 @@ private void parseJabRefComment(Map<String, String> meta) {
...
+    }
+
+    /**
+     * Parses comment types found in BibDesk, to migrate BibDesk Static Groups to JabRef.
+     * */
+    private void parseBibDeskComment(String comment, Map<String, String> meta) {
+        String xml = comment.substring(MetaData.BIBDESK_STATIC_FLAG.length() + 1, comment.length() - 1);
+        try {
+            // Build a document to handle the xml tags
+            DocumentBuilder builder = DocumentBuilderFactory.newInstance().newDocumentBuilder();
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/importer/fileformat/BibtexParser.java:384]: Extract this factor creation as a property variable because creating the factory is an expensive task

File: src/main/java/org/jabref/logic/importer/fileformat/BibtexParser.java [hunk 4]
```diff
@@ -330,6 +349,70 @@ private void parseJabRefComment(Map<String, String> meta) {
...
+            Document doc = builder.parse(new ByteArrayInputStream(xml.getBytes()));
+            doc.getDocumentElement().normalize();
+
+            NodeList dictList = doc.getElementsByTagName("dict");
+            meta.putIfAbsent("databaseType", "bibtex;");
+            meta.putIfAbsent("grouping", "0 AllEntriesGroup:;");
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/importer/fileformat/BibtexParser.java:390]: Use the constants from Metadata

File: src/main/java/org/jabref/logic/importer/fileformat/BibtexParser.java [hunk 5]
```diff
@@ -330,6 +349,70 @@ private void parseJabRefComment(Map<String, String> meta) {
...
+
+                String groupName = null;
+                String citationKeys = null;
+
+                // Retrieves group name and group entries and adds these to the metadata
+                for (int j = 0; j < keyList.getLength(); j++) {
+                    if (keyList.item(j).getTextContent().matches("group name")) {
+                        groupName = stringList.item(j).getTextContent();
+                        String oldValue = meta.get("grouping"); // if there are multiple groups we concatinate to the existing String
+                        meta.put("grouping", oldValue + "1 StaticGroup:" + groupName + "\\;0\\;0\\;\\;\\;\\;;");
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/importer/fileformat/BibtexParser.java:406]: Not sure how useful this is but you could alternatively try to use the groupsParser + serializer

https://github.com/JabRef/jabref/blob/004e2813368314b22821413d268b13a227af0a9d/src/test/java/org/jabref/logic/exporter/GroupSerializerTest.java#L52

File: src/main/java/org/jabref/logic/importer/fileformat/BibtexParser.java [hunk 6]
```diff
@@ -330,6 +349,70 @@ private void parseJabRefComment(Map<String, String> meta) {
...
+                    } else if (keyList.item(j).getTextContent().matches("keys")) {
+                        citationKeys = stringList.item(j).getTextContent(); // adds group entries
+                    }
+                }
+                // Adds the group name and citation keys to the field so all the entries can be added in the groups once parsed
+                parsedBibdeskGroups.put(groupName, citationKeys);
+            }
+        } catch (ParserConfigurationException | IOException | SAXException e) {
+            throw new RuntimeException(e);
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/importer/fileformat/BibtexParser.java:415]: Just add a logg here, we don't want to crash the whole application

File: src/main/java/org/jabref/logic/importer/fileformat/BibtexParser.java [hunk 7]
```diff
@@ -624,7 +707,23 @@ private void parseField(BibEntry entry) throws IOException {
...
+                    try {
+                        // Parse the base64 encoded binary plist to get the relative (to the .bib file) path
+                        NSDictionary plist = (NSDictionary) BinaryPropertyListParser.parse(decodedBytes);
+                        NSString relativePath = (NSString) plist.objectForKey("relativePath");
+                        Path path = Path.of(relativePath.getContent());
+
+                        LinkedFile file = new LinkedFile("", path, "");
+                        entry.addFile(file);
+                    } catch (Exception e) {
+                        throw new IOException();
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/importer/fileformat/BibtexParser.java:722]: Add a message to the exception e.

**Review by Siedlerchr — CHANGES_REQUESTED** [review]
Thank you very much! In general this already looks good, I only have a few smaller comments

File: external-libraries.md [hunk 8]
```diff
@@ -133,6 +133,20 @@ URL:     https://github.com/weisJ/jsvg
...
+License: Apache-2.0
+```
+
+```yaml
+Id:      com.googlecode.plist:dd-plist
+Project: com.dd.plist - A Java library for working with property lists
+URL:     https://github.com/3breadt/dd-plist
+License: MIT
+```
+
```
Comment (calixtus) [inline external-libraries.md:149]: Do not edit external-libraries by hand. It will be edited when releasing a new version. There are too many possible errors. Please revert the changes to this file.

File: src/main/java/org/jabref/logic/importer/fileformat/BibtexParser.java [hunk 9]
```diff
@@ -318,7 +337,7 @@ private void parseJabRefComment(Map<String, String> meta) {
                 }
             }
         } else if (comment.substring(0, Math.min(comment.length(), MetaData.ENTRYTYPE_FLAG.length()))
-                          .equals(MetaData.ENTRYTYPE_FLAG)) {
+                .equals(MetaData.ENTRYTYPE_FLAG)) {
```
Comment (calixtus) [inline src/main/java/org/jabref/logic/importer/fileformat/BibtexParser.java:340]: please do not change the formatted indentation.

File: external-libraries.md (hunk 8 above)
Comment (Siedlerchr) [reply external-libraries.md:149]: Na actually it looks good so far

Comment (koppor) [reply external-libraries.md:149]: The concrete change is OK for me. However, the changes below are not.

`commons-collections:commons-collections:3.2.2` is added, but no added entry in `external-libraries.md`. Thus, either all or none.

File: licenses/com.googlecode.plist_ddplist.txt [hunk 10]
```diff
@@ -0,0 +1,20 @@
+dd-plist - An open source library to parse and generate property lists
```
Comment (koppor) [inline licenses/com.googlecode.plist_ddplist.txt:1]: Haha, I totally forgot about the `licenses` sub folder... Lat change: 9 years ago...

OK to add something there... We need to cleanup at some stage.

SBOM generation etc.

File: src/main/java/org/jabref/logic/importer/fileformat/BibtexParser.java [hunk 11]
```diff
@@ -227,6 +244,8 @@ private ParserResult parseFileContent() throws IOException {
             skipWhitespace();
         }
 
+        addBibDeskGroupEntriesToJabRefGroups();
```
Comment (koppor) [inline src/main/java/org/jabref/logic/importer/fileformat/BibtexParser.java:247]: Can this method be moved to a separate class?

Reason: It seems that `private Map<String, String> parsedBibdeskGroups;` is a temporary data and the data from BibDesk is moved to JabRef's internal data model.

The parsing logic is very large and maybe, we can split it up a bit.

File: src/main/java/org/jabref/logic/importer/fileformat/BibtexParser.java (hunk 5 above)
Comment (koppor) [reply src/main/java/org/jabref/logic/importer/fileformat/BibtexParser.java:406]: Yeah, this is far too low level. There is a JabRef data structure for groups, which should be used.

File: src/test/java/org/jabref/logic/importer/fileformat/BibtexParserTest.java [hunk 12]
```diff
@@ -1863,4 +2164,26 @@ void parseDuplicateKeywordsWithTwoEntries() throws Exception {
...
+                    url = {https://arxiv.org/pdf/2309.06478.pdf},
+                    year = {2023},
+                    bdsk-file-1 = {YnBsaXN0MDDSAQIDBFxyZWxhdGl2ZVBhdGhZYWxpYXNEYXRhXxAeLi4vLi4vRG93bmxvYWRzLzIzMDkuMDY0NzgucGRmTxEBUgAAAAABUgACAAAMTWFjaW50b3NoIEhEAAAAAAAAAAAAAAAAAAAA4O/yLkJEAAH/////DjIzMDkuMDY0NzgucGRmAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAP/////hKRkeAAAAAAAAAAAAAgACAAAKIGN1AAAAAAAAAAAAAAAAAAlEb3dubG9hZHMAAAIAKy86VXNlcnM6Y2hyaXN0b3BoczpEb3dubG9hZHM6MjMwOS4wNjQ3OC5wZGYAAA4AHgAOADIAMwAwADkALgAwADYANAA3ADgALgBwAGQAZgAPABoADABNAGEAYwBpAG4AdABvAHMAaAAgAEgARAASAClVc2Vycy9jaHJpc3RvcGhzL0Rvd25sb2Fkcy8yMzA5LjA2NDc4LnBkZgAAEwABLwAAFQACABH//wAAAAgADQAaACQARQAAAAAAAAIBAAAAAAAAAAUAAAAAAAAAAAAAAAAAAAGb},
+                    bdsk-url-1 = {https://arxiv.org/abs/2309.06478}}
+                    }
+                """));
+        BibDatabase database = result.getDatabase();
+        BibEntry entry = database.getEntryByCitationKey("Kovakkuni:2023aa").get();
+        String link = entry.getFiles().getFirst().getLink();
+        assertEquals(link, "../../Downloads/2309.06478.pdf");
```
Comment (koppor) [inline src/test/java/org/jabref/logic/importer/fileformat/BibtexParserTest.java:2187]: It's hard to see the effects of the import. Is `bdsk-file-1` be kept?

Maybe, you can test with

```java
BibEntry expected = new BibEntry(StandardType.Article)
  .withField(...)
  .withField(...)
   ...
assertEquals(List.of(expected), result.getDatabase().getEntries());
```

**Review by koppor — CHANGES_REQUESTED** [review]
Thank you for working on this. The code works, but we also have "good code qualtiy" in "definition-of-done", thus some comments in that direction.

File: external-libraries.md (hunk 8 above)
Comment (calixtus) [reply external-libraries.md:149]: So lets keep just the license and revert the other changes. Have to be done anyway on next release. That what I call in german "verschwendete Liebesmüh".

**Review by koppor — DISMISSED** [review]

**Review by calixtus — APPROVED** [review]

**Comment by Siedlerchr** [comment]
TODO: Just a quick thougt I had this morning, we need to not only set Groups but we need to mege the existing group tree from metadata. Because BIbDesk is not an importer it's still a bib file

File: src/main/java/org/jabref/logic/bibtex/comparator/BibDatabaseDiff.java [hunk 13]
```diff
@@ -89,6 +89,10 @@ private static List<BibEntryDiff> compareEntries(List<BibEntry> originalEntries,
                 }
             }
 
+            // Prevent IndexOutOfBoundException
+            if (newEntries.isEmpty()) {
+                return differences;
+            }
```
Comment (koppor) [inline src/main/java/org/jabref/logic/bibtex/comparator/BibDatabaseDiff.java:95]: Shouldn't that go up to the beginning of the method? differences is not filled up to now, newEntries is also not modified.

**Review by Siedlerchr — APPROVED** [review]
